protocols/mdns: Allow users to choose between async-io and tokio runtime

gallegogt opened the pull request:
**Problem**

When use the Tokio library with mdns module, it is polling every time the UDP socket; causing hight CPU usage.

**Solution**

When using Tokio library, use the UppSocket provided by the Tokio library instead of the Async-io wrapper.

**Breaking Change**

After this fix exist two features ```async-io```, ```tokio``` for this module, for the general module must be use the features ```mdns-async-io``` or ```mdns-tokio``` for async-io or tokio libraries

## Links to any relevant issues

* https://github.com/libp2p/rust-libp2p/issues/2675
* https://github.com/libp2p/rust-libp2p/issues/2591

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

gallegogt committed:
test(mdns): renamed mdns tests

gallegogt committed:
fix(protocols/mdns): fix high cpu usage with tokio library

**Problem**

When use the Tokio library with mdns module, it is polling every time the UDP socket; causing hight CPU usage.

**Solution**

When using Tokio library, use the UppSocket provided by the Tokio library instead of the Async-io wrapper.

**Breaking Change**

After this fix exist two features ```async-io```, ```tokio``` for this module, for the general module must be use the features ```mdns-async-io``` or ```mdns-tokio``` for async-io or tokio libraries

gallegogt committed:
fix(examples/chat-tokio):update mdns-tokio feature

gallegogt committed:
chore(protocols/mdns): update changelog

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 - Update to `libp2p-swarm` `v0.38.0`.
 
+- fix(protocols/mdns): fix high cpu usage with tokio library.
```
Please follow the existing format of the changelog entries! :)

thomaseizinger commented on protocols/mdns/src/behaviour/socket.rs:
```diff
@@ -0,0 +1,201 @@
+};
+
+/// Interface that must be implemented by the different runtimes to use the UdpSocket in async mode
+///
```
```suggestion
/// Interface that must be implemented by the different runtimes to use the UdpSocket in async mode.
```

I don't think these newlines are idiomatic rustdoc.

thomaseizinger commented on protocols/mdns/src/behaviour.rs:
```diff
@@ -46,7 +56,11 @@ pub struct Mdns {
+    iface_states: HashMap<IpAddr, InterfaceState<AsyncUdpSocket, AsyncTimer>>,
+
+    #[cfg(feature = "tokio")]
+    iface_states: HashMap<IpAddr, InterfaceState<TokioUdpSocket, AsyncTimer>>,
```
Features should be additive in Rust but this code will break at compile-time if both are enabled.

I think we might need to go with an approach similar as in `libp2p-tcp` here: Create a dedicated `TokioMdns` and `AsyncIoMdns` struct and enable them conditionally.

thomaseizinger reviewed (changes requested):
Thanks!

I think the feature itself is worthwhile having but we need to implement it slightly different. Commands like `cargo build --all-features` should continue to work across the workspace.

gallegogt committed:
fix(protocols/mdns):permit build with all features

**Breaking Changes**

After this change must be used `Mdns` or `TokioMdns`, depending on the enabled features

gallegogt committed:
chore(protocols/mdns): update changelog

thomaseizinger commented on protocols/mdns/Cargo.toml:
```diff
@@ -25,8 +24,26 @@ smallvec = "1.6.1"
+[features]
+default = ["async-io"]
+tokio = ["tokio-crate"]
+async-io = ["async-io-crate"]
```
I think it would be good to use the `dep:` syntax here to hide the other two features.

gallegogt replied:
@thomaseizinger  Do you mean use `tokio = ["dep:tokio"]` instead of `tokio = ["tokio-crate"]`?

thomaseizinger commented on protocols/mdns/src/behaviour.rs:
```diff
@@ -35,18 +39,36 @@ use smallvec::SmallVec;
+pub struct GenMdns<S, T>
+where
+    T: TimerBuilder,
+    S: AsyncSocket,
```
I think we can remove these and add them only at the impl block where they are needed.

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -34,20 +35,24 @@ use std::{
+pub struct InterfaceState<U, T>
+where
+    U: AsyncSocket,
+    T: TimerBuilder,
```
Same here.

thomaseizinger commented on protocols/mdns/src/behaviour/socket.rs:
```diff
@@ -0,0 +1,190 @@
+        _buf: &mut [u8],
+    ) -> Poll<Option<(usize, SocketAddr)>> {
+        Poll::Pending
+    }
```
Not sure about these default implementations? Why have them?

gallegogt replied:
I will remove that default implementations.

thomaseizinger commented on protocols/mdns/src/behaviour/timer.rs:
```diff
@@ -0,0 +1,131 @@
+    fn interval_at(start: Instant, duration: Duration) -> Self;
+
+    /// Poll the timer
+    fn poll_tick(&mut self, cx: &mut Context) -> Poll<Self::Item>;
```
Instead of this fn here, what about implementing `Stream`?

This would make the trait naming better as this is the only none-builder function atm.

thomaseizinger commented on examples/chat-tokio.rs:
```diff
@@ -33,15 +33,19 @@
-    mdns::{Mdns, MdnsEvent},
+    mdns::{
+        // `MdnsTokio` is available through the `mdns-tokio` feature.
+        TokioMdns, 
```
Typo.


```suggestion
        // `TokioMdns` is available through the `mdns-tokio` feature.
        TokioMdns, 
```

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 - Update to `libp2p-swarm` `v0.38.0`.
 
+- Fix high CPU usage with Tokio library (see [PR 2748]).
```
Missing link to PR.

thomaseizinger commented on protocols/mdns/src/behaviour.rs:
```diff
@@ -19,10 +19,14 @@
 use self::iface::InterfaceState;
+
+use crate::behaviour::{socket::AsyncSocket, timer::TimerBuilder};
+
```
Why these newlines?

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -24,8 +24,9 @@ mod query;
-use futures::prelude::*;
+
+use crate::behaviour::{socket::AsyncSocket, timer::TimerBuilder};
+
```
These newlines are unnecessary too.

thomaseizinger commented on protocols/mdns/src/behaviour/socket.rs:
```diff
@@ -0,0 +1,190 @@
+            packet: &[u8],
+            to: SocketAddr,
+        ) -> Poll<()> {
+            if self.poll_writable(cx).is_ready() {
```
Nit: Could use the `ready` macro here (and in a few other places).

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -198,43 +205,40 @@ impl InterfaceState {
+                Poll::Ready(_) => log::trace!("sent packet on iface {}", self.addr),
+                Poll::Pending => s_buffer.push_front(packet),
             }
         }
```
Couldn't we already break from the loop on the first `Pending`?

Seems wasteful to empty the entire send buffer.

thomaseizinger commented on protocols/mdns/src/behaviour/socket.rs:
```diff
@@ -0,0 +1,190 @@
+                    }
+                },
+                Poll::Ready(Err(e)) => {
+                    log::error!("Failed recv ready datagram: {}", e);
```
These error messages have copy paste errors.

thomaseizinger commented on protocols/mdns/src/behaviour/socket.rs:
```diff
@@ -0,0 +1,190 @@
+                    }
+                    Err(err) => {
+                        log::error!("Failed reading datagram: {}", err);
+                        return Poll::Ready(());
```
I am not sure about returning `Ready` here. Why not propagate the error?

thomaseizinger reviewed (changes requested):
Thanks!

I gave this a more in-depth read, please see the inline comments :)

gallegogt committed:
style(chat-tokio): fix formating error

gallegogt committed:
chore(protocols/mdns): use deps syntax

gallegogt committed:
fix(protocols/mdns): clean the code

gallegogt committed:
fix(examples/chat-tokio): fix typo

gallegogt committed:
chore(protocols/mdns): add pr link to changelog

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 - Update to `libp2p-swarm` `v0.38.0`.
 
+- Fix high CPU usage with Tokio library (see [PR 2748](https://github.com/libp2p/rust-libp2p/pull/2748) ).
```
```suggestion
- Fix high CPU usage with Tokio library. See [PR 2748].

[PR 2748]: https://github.com/libp2p/rust-libp2p/pull/2748
```

Typically, we put the link at the bottom.

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -69,7 +68,11 @@ pub struct InterfaceState {
+impl<U, T> InterfaceState<U, T>
+where
+    U: AsyncSocket,
+    T: TimerBuilder + futures::Stream,
```
I am surprised that we can say `futures::Stream` here without specifying the `Item` type?

gallegogt replied:
In this case, the `Stream::Item` is not used, so it is not required to check it at compile time.

thomaseizinger commented on protocols/mdns/src/behaviour/socket.rs:
````diff
@@ -0,0 +1,149 @@
+/// Interface that must be implemented by the different runtimes to use the UdpSocket in async mode
+pub trait AsyncSocket: Send + 'static {
+    /// Create the async socket from the ```std::net::UdpSocket```
+    fn from_socket(socket: UdpSocket) -> std::io::Result<Self>
````
```suggestion
    fn from_std(socket: UdpSocket) -> std::io::Result<Self>
```

Just an idea, not attached to it.

gallegogt replied:
@thomaseizinger I'll check it but, but the Tokio socket implementation have a function with this name, and might cause conflict

thomaseizinger replied:
If we only import the `AsyncSocket` type at the usage site, it should be unambiguous.

thomaseizinger commented on protocols/mdns/src/behaviour/socket.rs:
```diff
@@ -0,0 +1,149 @@
+        _cx: &mut Context,
+        _packet: &[u8],
+        _to: SocketAddr,
+    ) -> Poll<Result<(), Error>>;
```
```suggestion
    /// Attempts to receive a single packet on the socket from the remote address to which it is connected.
    fn poll_read(
        &mut self,
        _cx: &mut Context,
        _buf: &mut [u8],
    ) -> Poll<Result<Option<(usize, SocketAddr)>, Error>>;

    /// Attempts to send data on the socket to a given address.
    fn poll_write(
        &mut self,
        _cx: &mut Context,
        _packet: &[u8],
        _to: SocketAddr,
    ) -> Poll<Result<(), Error>>;
```

`poll_read` and `poll_write` are more common I think but at the same time, this might be a bit ambiguous at the usage site.

gallegogt replied:
@thomaseizinger the same with this function, the Tokio socket implementation have a function with this name, and might cause conflict

thomaseizinger commented on protocols/mdns/src/behaviour/socket.rs:
```diff
@@ -0,0 +1,149 @@
+        &mut self,
+        _cx: &mut Context,
+        _buf: &mut [u8],
+    ) -> Poll<Result<Option<(usize, SocketAddr)>, Error>>;
```
Can we add some documentation here in what happens to the buffer in case of `Poll::Pending`? Does the caller have a guarantee that the buffer will only be touched if this returns `Poll::Ready`?

thomaseizinger commented on protocols/mdns/src/behaviour/socket.rs:
```diff
@@ -0,0 +1,149 @@
+                Poll::Ready(Ok(_)) => match self.try_recv_from(buf) {
+                    Ok((len, from)) => Poll::Ready(Ok(Some((len, from)))),
+                    Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                        Poll::Ready(Ok(None))
```
The [docs](https://docs.rs/tokio/1.20.0/tokio/net/struct.UdpSocket.html#errors-2) say this can never happen so I think we could `panic!` here? That should allow us to remove the `Option` type from the return value if I am not mistaken.

gallegogt replied:
In Tokio case, from documentation.

[Tokio try_recv_from](https://docs.rs/tokio/1.20.0/tokio/net/struct.UdpSocket.html#method.try_recv_from)

```
// Try to recv data, this may still fail with `WouldBlock`
// if the readiness event is a false positive.
```

Throw panic when is WouldBlock, does not seem to be a good option in this case.

thomaseizinger replied:
Ah you are right, I was mis-reading the code!

What we can then do is put all of this in a loop I think. In case we encounter `WouldBlock`, we can loop up to `poll_recv_ready` which should then return `Pending` and register a Waker. This should allow us to get rid of the `Option` return value!

Thoughts?

thomaseizinger replied:
The other alternative is to also just forward the error to our user. This has likely the advantage that we only have to write the loop once.

I think this is better because even with the loop here, we can't remove the `WouldBlock` case from the signature so any caller may still think it has to handle it.

thomaseizinger commented on protocols/mdns/src/behaviour/socket.rs:
```diff
@@ -0,0 +1,149 @@
+            match self.poll_send_ready(cx) {
+                Poll::Ready(Ok(_)) => match self.try_send_to(packet, to) {
+                    Ok(_len) => Poll::Ready(Ok(())),
+                    Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => Poll::Ready(Ok(())),
```
The [example](https://docs.rs/tokio/1.20.0/tokio/net/struct.UdpSocket.html#example-6) uses a loop here to avoid the `WouldBlock` case. I think we should do that too.

thomaseizinger commented on protocols/mdns/src/behaviour/socket.rs:
```diff
@@ -0,0 +1,149 @@
+            cx: &mut Context,
+            buf: &mut [u8],
+        ) -> Poll<Result<Option<(usize, SocketAddr)>, Error>> {
+            match self.poll_recv_ready(cx) {
```
You can use `?` here and also the `ready!` macro if you want.

```suggestion
            match futures::ready!(self.poll_recv_ready(cx)?) {
```

thomaseizinger commented on protocols/mdns/src/behaviour/timer.rs:
```diff
@@ -0,0 +1,129 @@
+
+/// Simple wrapper for the differents type of timers
+#[derive(Debug)]
+pub struct WrapTimer<T> {
```
Why not just call this `Timer`?

thomaseizinger commented on protocols/mdns/src/behaviour/timer.rs:
```diff
@@ -0,0 +1,129 @@
+}
+
+/// Builder interface to homogenize the differents implementations
+pub trait TimerBuilder: Send + Unpin + 'static {
```
Could name this `Builder` and refer to it as `timer::Builder` if we want the additional context.

thomaseizinger commented on protocols/mdns/src/behaviour/timer.rs:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    impl Stream for AsyncTimer {
+        type Item = Instant;
```
Do we ever need the item type? Otherwise we could map it to `()` and thus have the same signature for both modules.

gallegogt replied:
I would prefer to leave it as it is, in case the instant is required later

thomaseizinger commented on protocols/mdns/src/behaviour/timer.rs:
```diff
@@ -0,0 +1,129 @@
+            }
+        }
+
+        /// Creates a timer that emits events periodically, starting at start.
```
I don't think we need to copy the docs from the trait definition over.

thomaseizinger commented on protocols/mdns/tests/use-tokio.rs:
```diff
@@ -0,0 +1,137 @@
+use std::error::Error;
+use std::time::Duration;
+
+async fn create_swarm(config: MdnsConfig) -> Result<Swarm<TokioMdns>, Box<dyn Error>> {
```
I'd prefer the tests to be at the top and the utility functions to be at the bottom of the file.

thomaseizinger commented on src/lib.rs:
```diff
@@ -79,8 +79,7 @@ pub use libp2p_identify as identify;
 #[doc(inline)]
 pub use libp2p_kad as kad;
-#[cfg(feature = "mdns")]
-#[cfg_attr(docsrs, doc(cfg(feature = "mdns")))]
```
Seems like this line with the `doc(cfg)` got lost!

thomaseizinger reviewed (commented):
Thanks!

This looks really good now, I've left a couple more comments :)

gallegogt committed:
chore(protocols/mdns): fix changelog

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -198,43 +199,49 @@ impl InterfaceState {
-                    }
+        match self
+            .recv_socket
+            .poll_receive_packet(cx, &mut self.recv_buffer)
```
I think this should be a loop that reads for as long as it returns `Poll::Ready`.

gallegogt replied:
@thomaseizinger  Is the buffer too long to require a loop?

gallegogt committed:
fix(protocols/mdns): rename socket functions

gallegogt committed:
fix(protocols/mdns): rename TimerBuilder trait

gallegogt committed:
fix(protocols/mdns): add cfg_attr docsr

gallegogt committed:
fix(protocols/mdns): update mdns tests

gallegogt committed:
chore(protocols/mdns): add fmt

gallegogt committed:
fix(protocols/mdns): update socket and timer impl

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -199,37 +199,39 @@ where
-                    self.inject_mdns_packet(packet, params);
+        match self.recv_socket.poll_read(cx, &mut self.recv_buffer) {
+            Poll::Ready(result) => match result {
+                Ok((len, from)) => {
```
I'd prefer to not have nested `match` expressions but instead, nest the patterns as the original implementation did.

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -107,7 +110,9 @@ impl InterfaceState {
         };
-        let send_socket = Async::new(UdpSocket::bind(bind_addr)?)?;
+        let std_socket = UdpSocket::bind(bind_addr)?;
+        let send_socket = U::from_std(std_socket)?;
```
For the sake of minimizing the diff, we could write this similar to how the old code did it:

```suggestion
        let send_socket = U::from_std(UdpSocket::bind(bind_addr)?)?;
```

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -171,17 +175,14 @@ impl InterfaceState {
-                    for addr in addrs {
-                        self.discovered
-                            .push_back((*peer.id(), addr, new_expiration));
+                        self.discovered.push_back((*peer.id(), da, new_expiration));
```
This is not functionally equivalent I think. Previously, we've pushed the translated AND untranslated address whereas now we only push one or the other, depending on whether the translation worked. What is the reason for the behaviour change?

gallegogt replied:
Sorry, in this case reduce the 2 loops to 1:

```
for addr in peer.addresses() {                                                      
     if let Some(new_addr) = address_translation(addr, &observed) {                  
        self.discovered.push_back((*peer.id(), new_addr.clone(), new_expiration));  
     }                                                                               
                                                                                           
     self.discovered.push_back((*peer.id(), addr.clone(), new_expiration));                  
 } 
```

thomaseizinger replied:
It is fine to reduce the two loops to one but we need to keep the same functionality :)

The code snipped you posted seems good!

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -198,43 +199,51 @@ impl InterfaceState {
-                .now_or_never()
-            {
-                Some(Ok((len, from))) => {
+        match self.recv_socket.poll_read(cx, &mut self.recv_buffer) {
```
This should be a loop similar as to before.

Previously, we read as many packets as possible until we hit `Poll::Pending`. Now we are reading at most one packet and then moving on to sending responses. We should exhaust the `recv_socket` until it returns `Pending`.

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -198,43 +199,51 @@ impl InterfaceState {
+                    Ok(_) => log::trace!("sent packet on iface {}", self.addr),
+                    Err(err) => {
+                        log::error!("error sending packet on iface {} {}", self.addr, err);
+                        self.send_buffer.push_front(packet);
```
This is not functionally equivalent to what we did before.

Previously, we discarded a packet that we failed to send AND moved on to the next one if the socket was still ready to accept packets.

Now we are pushing the packet back into the buffer and breaking from the loop.

thomaseizinger reviewed (changes requested):
Thank you for incorporating the feedback :)

I think this is close to being mergeable. I made a few comments where we don't seem to be doing quite the same as previously. Please comment if that is on purpose, otherwise I'd prefer to stick to the previous logic.

gallegogt committed:
fix(protocols/mdns): fix implementation logic and style

mxinden commented on examples/chat-tokio.rs:
```diff
@@ -33,15 +33,19 @@
 //!cargo run --example chat-tokio \\
 //!    --no-default-features \\
-//!    --features="floodsub mplex noise tcp-tokio mdns"
+//!    --features="floodsub mplex noise tcp-tokio mdns-tokio"
```
:pray:

mxinden commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -2,8 +2,12 @@
 
 - Update to `libp2p-swarm` `v0.38.0`.
 
+- Fix high CPU usage with Tokio library (see [PR 2748]).
```
Can you document the added feature flags here? What should an `async-std` and `tokio` user do when upgrading to the new version?

mxinden commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -198,43 +201,52 @@ impl InterfaceState {
-                Some(Err(err)) => log::error!("Failed reading datagram: {}", err),
-                None => {}
+                Poll::Ready(Err(err)) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                    // Not more bytes available on the socket to read
```
```suggestion
                    // No more bytes available on the socket to read.
```

Not a native speaker. Might be wrong.

mxinden commented on protocols/mdns/src/behaviour/socket.rs:
```diff
@@ -0,0 +1,129 @@
+    task::{Context, Poll},
+};
+
+/// Interface that must be implemented by the different runtimes to use the UdpSocket in async mode
```
```suggestion
/// Interface that must be implemented by the different runtimes to use the [`UdpSocket`] in async mode
```

mxinden commented on protocols/mdns/src/behaviour/socket.rs:
````diff
@@ -0,0 +1,129 @@
+
+/// Interface that must be implemented by the different runtimes to use the UdpSocket in async mode
+pub trait AsyncSocket: Send + 'static {
+    /// Create the async socket from the ```std::net::UdpSocket```
````
```suggestion
    /// Create the async socket from the [`std::net::UdpSocket`].
```

mxinden reviewed (commented):
This looks good to me. Deferring final review to @thomaseizinger.

Thanks @gallegogt and @thomaseizinger for the solid work here!

gallegogt committed:
doc(protocols/mdns): fix documentation

thomaseizinger commented on protocols/mdns/CHANGELOG.md:
```diff
@@ -2,8 +2,14 @@
 
+- Allow users to choose between async-io and tokio runtime
+  in the mdns protocol implementation. `async-io` is a default
+  feature, with an additional `tokyo` feature  (see [PR 2748])
```
```suggestion
  feature, with an additional `tokio` feature  (see [PR 2748])
```

Small typo :)

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -198,43 +201,52 @@ impl InterfaceState {
-            {
-                Some(Ok((len, from))) => {
+        loop {
+            match self.recv_socket.poll_read(cx, &mut self.recv_buffer) {
```
Feel free to leave this as is but this could be written as a `while let` which would avoid the `Poll::Pending => break` clause.

```rust
while let Poll::Ready(result) = self.recv_socket.poll_read(cx, &mut self.recv_buffer) {
    match result {
        // ..
    }
}
```

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -198,43 +201,52 @@ impl InterfaceState {
+                Poll::Ready(Err(err)) => {
+                    log::error!("failed reading datagram: {}", err);
+                    break;
+                }
```
In the previous implementation, we did not `break` on an error. Please keep it that way.

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -198,43 +201,52 @@ impl InterfaceState {
+        loop {
+            match self.recv_socket.poll_read(cx, &mut self.recv_buffer) {
+                Poll::Ready(Ok((len, from))) => {
                     if let Some(packet) = MdnsPacket::new_from_bytes(&self.recv_buffer[..len], from)
```
An idea for a follow-up: We could create a wrapper that utilizes `AsyncSocket` to directly read and write `MdnsPacket`s, i.e. have a layer between this code here and the actual socket.

thomaseizinger commented on protocols/mdns/src/behaviour/iface.rs:
```diff
@@ -198,43 +201,52 @@ impl InterfaceState {
+                Poll::Pending => {
+                    self.send_buffer.push_front(packet);
+                    break;
                 }
```
Here too, we are `break`ing but previously didn't.

thomaseizinger reviewed (changes requested):
Thanks!

I've found too more misalignments with the previous version. Thanks for sticking to this :)

gallegogt committed:
fix(protocols/mdns): fix typo and logic

thomaseizinger reviewed (approved):
Thanks for implementing all the feedback!

This is ready now :)

gallegogt committed:
fix(protocols/mdns): sockets pinned on poll

gallegogt committed:
fix(protocols/mdns): change tokio interval missed tick behavior

gallegogt committed:
refactor(mdns/test): update tokio expiration test

gallegogt committed:
ci(*): testing workflow

gallegogt committed:
ci(*): update workflow name

gallegogt committed:
fix(protocol/mdns): change interval behavior for tokio timer impl

thomaseizinger commented on .github/workflows/testing.yml:
```diff
@@ -0,0 +1,134 @@
+name: Pre testing
```
Are you intending to merge this workflow as well or is this a leftover from testing?

gallegogt replied:
sorry it's a remnant of the tests, I'll delete it

thomaseizinger reviewed (approved):
Thanks a lot for this contribution! Great to see better support for multiple runtimes!

One last question regarding the workflow, otherwise this is good to be merged from my perspective!

gallegogt committed:
chore(workflow): remove testing workflow

gallegogt committed:
Permit use mdns with Tokio or AsyncIO runtime

mxinden committed:
protocols/mdns: Update changelog

Given the latest release, this patch now targets `libp2p-mdns` `v0.40.0` and not
`v0.39.0`.

mxinden reviewed (approved):
:pray: Thanks @gallegogt and @thomaseizinger!
